Slim/Control/Commands.pm

```diff
@@ -2050,7 +2050,7 @@ sub playlistcontrolCommand {
 			$what->{'year.id'} = $year_id;
 		}
 		
-		@tracks = _playlistXtracksCommand_parseSearchTerms($client, $what, $cmd);
+		@tracks = _playlistXtracksCommand_parseSearchTerms($client, $what, $cmd, $request->getParam('sort'));
```

Comment:
Wouldn't the `sort` value be part of the `$what` parameter?

Reply:
I'm not sure what you mean. As in set the **whole** `album.year, album.title, me.disc, me.track` sort setting here? I thought about that - but it would be duplicating code. Or do you mean pass the `sort:yearalbum` somehow in `$what`

Reply:
pass the `sort:...` parameter in `$what`. A the code is expecting it to be there anyway: it's already looking for that parameter.

Reply:
Line 3320(ish) (`while (my ($key, $value) = each %{$terms}) {`) iterates over all `key:value` pairs in the `$what` hash. If it finds `genre`, `album`, etc, it sets `$sort=$albumSort` If I set `sort:...` in `$what` then its usage will be order dependent. i.e. if the iteratation sees `sort` before `genre` it will get overwritten. Now, I assume they'll be iterated alphabetically so shold be OK - but that's just by chance.

Is the following roughly what you had in mind? (Assuming `sort:...` passed in `$what`)

```
(Line 3276-ish)
	my $albumSort = $albumSort . $sqlHelperClass->append0("album.titlesort") . " $collate"
		. ', me.disc, me.tracknum, '
		. $sqlHelperClass->append0("me.titlesort") . " $collate";
	my $albumYearSort = $sqlHelperClass->append0("album.year") . ", " . $albumSort;	
	my $trackSort = "me.disc, me.tracknum, " . $sqlHelperClass->append0("me.titlesort") . " $collate";

...
(Line 3391-ish)

		} elsif ($key eq 'sort') {

			if ($value eq 'artflow' || $value eq 'yearalbum' || $value eq 'yearartistalbum') {
				$sort = $albumYearSort;
			} else {
				$sort = $value;
			}

		} else {
```

Reply:
If that works, then yes, I'd think that should be it. I was wondering what would be sending other `$sort` values. Would have to be some kind of SQL snippet or something...

Reply:
Actually, having looked a bit closer this will not work - if they keys are iterated alphabetically. Both `genre` and `year` set `$sort=$albumSort` - so if `year` is in `$what` then `$sort=$albumYearSort` wlil be overwritten by `$sort=$albumSort`. I could work-around this by setting `$sort=$albumYearSort` for `year` as well - which I guess makes sense...

Reply:
So, just tried and I saw the following iteration order:
```
KEY:sort VAL:yearalbum
KEY:libraryTracks.library VAL:0
KEY:contributor.id VAL:89755
```
everse alphabetical? Either way, sort was set abnd then reset-set when `contributor.id` was spotted.

So, can;t really use `sort` in `$what`. Unless I take `sort` out of the iteration, and jsut check for it explicity afterwards? Therefore it would have precedence over others.

Reply:
AFAIK there's no well defined order iterating over a hash, and the behaviour changed between Perl versions...

Reply:
OK, so checking `sort` afterwards makse sense. In the changed version, I check for album sorts using year - so use `$albumYearSort` - then check that sort is not an album sort before using. I don't like the `elsif` part. Is there a better way to check if the `sort` string is a valid string? i.e. checking its an SQL sort, not an album sort?